File: mirror/tier2-php/showcase/sec_audit.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Security
// Module: sec_audit.php

class AuditLog {
    private $entries = [];
    
    public function logEvent(string $eventType, string $details): int {
        $this->entries[] = [
            'type' => $eventType,
            'details' => $details,
            'timestamp' => date('c'),
        ];
        return count($this->entries);
    }
    
    public function getCount(): int {
        return count($this->entries);
    }
    
    public function getEventsByType(string $eventType): array {
        return array_filter($this->entries, fn($e) => $e['type'] === $eventType);
    }
}

$audit = new AuditLog();
$audit->logEvent('startup', 'system initialized');
$audit->logEvent('login', 'user authenticated');
$audit->logEvent('access', 'admin access attempted');
$audit->logEvent('warning', 'multiple failed login attempts');

function check_security($audit, $operation) {
    $audit->logEvent('security_check', $operation);
    
    if ($operation === 'admin_access') {
        $audit->logEvent('warning', 'admin access attempted');
        return false;
    }
    
    return true;
}

check_security($audit, 'read_data');
check_security($audit, 'admin_access');

$result = [
    'total_events' => $audit->getCount(),
    'startup_events' => count($audit->getEventsByType('startup')),
    'warning_events' => count($audit->getEventsByType('warning')),
    'admin_blocked' => !check_security($audit, 'admin_access'),
    'auditing_active' => true,
];

echo json_encode($result);
?>

File: mirror/tier2-ruby/showcase/sec_audit.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Security
# Module: sec_audit.rb
# Purpose: Demonstrate security auditing

class AuditLog
  def initialize
    @entries = []
  end
  
  def log_event(event_type, details)
    entry = {
      type: event_type,
      details: details,
      timestamp: Time.now.iso8601,
    }
    @entries << entry
    @entries.size
  end
  
  def get_count
    @entries.size
  end
  
  def get_events_by_type(event_type)
    @entries.select { |e| e[:type] == event_type }
  end
end

audit = AuditLog.new
audit.log_event("startup", "system initialized")
audit.log_event("login", "user authenticated")
audit.log_event("access", "admin access attempted")
audit.log_event("warning", "multiple failed login attempts")

def check_security(audit, operation)
  audit.log_event("security_check", operation)
  
  if operation == "admin_access"
    audit.log_event("warning", "admin access attempted")
    return false
  end
  
  true
end

check_security(audit, "read_data")
check_security(audit, "admin_access")

result = {
  total_events: audit.get_count,
  startup_events: audit.get_events_by_type("startup").size,
  warning_events: audit.get_events_by_type("warning").size,
  admin_blocked: !check_security(audit, "admin_access"),
  auditing_active: true,
}

puts result.inspect
